### PR TITLE
Attention decoder Bahdanau-style

### PR DESCRIPTION
**Pull request by jindrahelcl**
The code in `decoding_function` should now follow the equations in [Bahdanau et al. (2015)](http://arxiv.org/abs/1409.0473)

API Changes:
- Added maxout layer to the decoder (its size can be specified by `maxout_size` config)
- Additional variables - experiments will be not compatible

### CODE REVIEW

File: neuralmonkey/decoding_function.py [hunk 1]
```diff
@@ -1,59 +1,93 @@
 import tensorflow as tf
 
-def attention_decoder(decoder_inputs, initial_state, attention_objects,
-                      embedding_size, cell, output_size=None,
-                      loop_function=None, dtype=tf.float32, scope=None):
+from neuralmonkey.logging import debug
 
-    if output_size is None:
-        output_size = cell.output_size
+def attention_decoder(decoder_inputs, initial_state, attention_objects,
```
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: Co to přejmenovat třeba na `rnn_decoding_loop`?

Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Klidně. Nechal bych ale v názvu attention. Taky bych nějak zkusil oddělit attentive decoder a normální. Ale asi někde jinde než tady.. Kvůli tý hidden vrstvě, kterou si to asi nechceš komplikovat, když tam ten attention nemáš. Vůbec je to taková featura navíc mi přijde. Co udělat novej dekodér s attentionama, kterej by podědil od toho obyčejnýho?

File: neuralmonkey/decoding_function.py [hunk 2]
```diff
@@ -1,59 +1,93 @@
 import tensorflow as tf
```
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: Docstring, k čemu je to dobré. 

Tohle vlastně slouží jenom RNN dekodéru. Pokud bude dekodér sequence labeling nebo nedej bože jednoho dne třeba CTC, tak, tak se to nepoužije. Zvážil bych jestli by se to nemělo přestěhovat do adresáře decoders nebo rovnou do `decoders/decoder.py`, který by se pak mohl jmenovat `rnn_decoder.py`.

File: neuralmonkey/decoding_function.py [hunk 3]
```diff
@@ -1,59 +1,93 @@
...
+                                        cell, maxout_size, output_size)
             outputs.append(output)
+            states.append(state)
 
     return outputs, states
 
+
+def decode_step(prev_output, prev_state, attention_objects, rnn_cell,
+                maxout_size, output_size):
+    ## compute c_i:
```
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: Odkazoval bych se na čísla rovnice a stránek v příslušných článcích.

Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Jenže oni to nemaj očíslovaný.. :-) Jedině napsat do docstringu že to implementuje rovnice ze sekce A.2.2 na stranách 13 a 14..

Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Ale dobrá připomínka. Začal jsem dělat docstringy a uvědomil jsem si že tam mam pořád jednu projekci navíc. Na konci maxout_projection se returnuje a nemá tam co dělat asi.

File: neuralmonkey/decoding_function.py [hunk 4]
```diff
@@ -1,59 +1,93 @@
...
+    return output, state
+
+
+def linear_projection(states, size):
+    with tf.variable_scope("AttnOutputProjection"):
+        output = tf.nn.seq2seq.linear(states, size, True)
+        return output
+
+
+def maxout_projection(prev_state, prev_output, current_contexts,
```
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: Nemělo by se to odstěhovat někam do `nn` jako maxout_layer? 

Úplně nejhezčí a nejobecnější by asi bylo, kdyby ta projekce byla instance multilayer-perceptronu, kterej by se ale musel zobecnit, aby:
* na konci nemusel mít softmax
* mohl mít kromě dense layers ještě pofiderní vrstvy jako je highway layer a max-out

To asi není důležitý, tak jestli se ti s tím nechce teď dělat, tak se to asi může napsat do nějaké issue do budoucna.

File: neuralmonkey/decoding_function.py [hunk 5]
```diff
@@ -1,59 +1,93 @@
...
+                      maxout_size, output_size):
+    # in Bahdanau: t_i = [max(t'_{i,2j-1}, t'_{i,2j})] ... j = 1,...,l
```
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: Do docstringu, že maxout je z tohodle článku: https://arxiv.org/pdf/1302.4389.pdf

File: neuralmonkey/decoding_function.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: A když tohle dostane prázdnej seznam attentionů nebude to stejný?

Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: podle mě tam bude navíc spousta zbytečnejch projekcí

File: neuralmonkey/decoding_function.py (hunk 3 above)
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: Napiš, že to implementuje rovnice z A.2.2, abych to příště zase nehledal v textu článku a pak  teprve zjistil, že to má apendix.

Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: už to tam mam..

File: neuralmonkey/decoding_function.py (hunk 1 above)
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: Hm, to asi jo. Ale dekodér bez attentionu je dneska skoro kuriozita, ne?

Tím chci říct: jestli se ti to chce udělat, tak směle do toho, je to určitě krok k většímu ordnungu, ale mně by se to moc dělat nechtělo.

Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: No třeba ten obrázkovej bez attentionu byl, že jo.. Ještě si rozmyslim, jestli je to potřeba jako součást tohodle PR nebo to jde pak dodělat dodatečně.

File: neuralmonkey/decoding_function.py (hunk 3 above)
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: zčekni to teďkon

File: neuralmonkey/decoding_function.py (hunk 2 above)
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: taky bych to chtěl dát k dekodéru. Nebude to pak ale moc hnusnej soubor?

Btw. co je ctc?

File: neuralmonkey/decoding_function.py (hunk 4 above)
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Zamyslim se nad tim. Spíš jo, ale musim zvážit následky.

File: neuralmonkey/decoding_function.py (hunk 5 above)
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Hotovo v 69f3d9cdd

File: neuralmonkey/decoding_function.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: CTC je tohle: http://www.cs.toronto.edu/~graves/icml_2006.pdf
Používalo se na to speech recognition a handwriting recognition.

Bude hnusnej, ale spraví se to pak, že se zase půlka vycestuje do toho bezattentionovýho dekodéru.

Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Asi bych si tenhle refactor schoval na jindy. Už ho máme podchycenej v #9

Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Docstring jsem dodal.

File: neuralmonkey/decoding_function.py (hunk 1 above)
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Zahrnul bych to do refactoru `decoding_function` v samostatný issue nebo v rámci #9

File: neuralmonkey/decoding_function.py (hunk 4 above)
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Přidal jsem modul `nn.projection`, kde jsou funkce `linear` a `maxout`. Ten linear v tensorflow vypadá, že bude deprecovanej, proto bych ho nechal u nás a kdyžtak udělal ručně. (Teď používá `tf.nn.seq2seq.linear` s biasama)

**Comment by jindrahelcl** [comment]
All comments have been addressed. :-)

File: neuralmonkey/decoding_function.py (hunk 1 above)
Comment (jindrahelcl) [inline neuralmonkey/decoding_function.py]: Btw. dekodér bez attentionu je třeba ten language model, ne?

File: neuralmonkey/decoding_function.py (hunk 2 above)
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: OK.

**Comment by jlibovicky** [comment]
And how does it perform on the data?

I also noticed it failed to query the web service. We should probably check in the test script whether the HTTP status was 200.

**Comment by jindrahelcl** [comment]
Yes, I've noticed that too some time ago. It seems the tests of the server started failing after the decoder refactor so I would not deal with that here. I will make an issue out of this.

**Comment by jindrahelcl** [comment]
As for the performance, it is quite similar to the previous runs. Here are screenshots from tensorboard - red one is bahdanau, blue one is the original neuralmonkey. The common parameters and data are the same.

tb_bleu
tb_losses
tb_train_losses

**Comment by jindrahelcl** [comment]
Finally, I've managed to run the training on twister2 with the paper's constants. We'll see now.

**Comment by jlibovicky** [comment]
On the same data as in the paper?

**Comment by jindrahelcl** [comment]
Yep.. they give the link to the selected data in a footnote in the paper.

File: neuralmonkey/nn/projection.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
+"""
+This module implements various types of projections.
+"""
+#tests: lint
+import tensorflow as tf
+
+def linear(inputs, size, scope="LinearProjection"):
```
Comment (jlibovicky) [inline neuralmonkey/nn/projection.py]: It seems to be almost the same as function `nn.mlp.dense`. We should keep only one of them, probably this one and add maybe add an optional non-linearity here.

File: neuralmonkey/decoding_function.py [hunk 7]
```diff
@@ -1,60 +1,87 @@
...
+            output, state = decode_step(decoder_input, state, attention_objects,
+                                        cell, maxout_size)
             outputs.append(output)
+            states.append(state)
 
     return outputs, states
 
+
+def decode_step(prev_output, prev_state, attention_objects,
+                rnn_cell, maxout_size):
```
Comment (jlibovicky) [inline neuralmonkey/decoding_function.py]: AFAI, the maxout projection is only in the particular Bahdanau's paper, so I would rather make the maxout optional. In ideal case, we should be able to specify how man layers of linear/maxout projection we want to make a which non-linearities should be applied on top of them.

File: neuralmonkey/nn/projection.py (hunk 6 above)
Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: addressed in f659020e33b9fba16090013d9c3f1d4e2ffcc6be

Comment (jindrahelcl) [inline neuralmonkey/nn/projection.py]: I would keep the non-linearity outside the call of this function. Or, create something like `nn.projection.nonlinear`, which would use `linear` and append the activation... But I think the code will be more readable with `tf.tanh(linear(tensor, size))` instead of `project(tensor, size, activation=tf.tanh)`

Comment (jlibovicky) [inline neuralmonkey/nn/projection.py]: OK

**Comment by jindrahelcl** [comment]
Takže jediný, co dodělat do týhle branche je konfigurovatelnej maxout a tvorba initial state? Ta by teda možná šla udělat už jinde.. (Nematus používá jako init state dekodéru průměr rnn stavů enkodéru místo projekce koncovýho stavu)

**Comment by jindrahelcl** [comment]
@jlibovicky Když už jsme u toho, co je vlastně default? Deep output bez skrytý vrstvy (Pascanu et al., 2014), tak, jak to používaj v Show, Attend and Tell (Xu et al., 2016)?

Čili jedna lineární projekce, kterou se akorát napasuje ten součet projekcí stavu, prev_output a kontextu(ů) na velikost slovníku?

**Comment by jlibovicky** [comment]
Udělal bych jako default projekci na velikost slovníku, ať se nemusí řešit, že je jinak velkej skrytej stav a jinak velkej slovník. Inital staty uděláme jinde, ať už je to všechno zmergovaný.

**Comment by jindrahelcl** [comment]
Lossy jsou teďkon z nějakýho důvodu vydělený stem.. Píšu to, abych na to nezapomněl. :-)
